projects/slap_manipulation/src/slap_manipulation/policy/voxel_grid.py

```diff
@@ -0,0 +1,222 @@
+from functools import reduce
+from operator import mul
+
+import torch
+from torch import nn
+
+MIN_DENOMINATOR = 1e-12
+INCLUDE_PER_VOXEL_COORD = False
+
+
+class VoxelGrid(nn.Module):
+    def __init__(
+        self,
+        coord_bounds,
+        voxel_size: int,
+        device,
+        batch_size,
+        feature_size,  # e.g. rgb or image features
+        max_num_coords: int,
+    ):
+        super(VoxelGrid, self).__init__()
+        self._device = device
+        self._voxel_size = voxel_size
+        self._voxel_shape = [voxel_size] * 3
+        self._voxel_d = float(self._voxel_shape[-1])
+        self._voxel_feature_size = 4 + feature_size
+        self._voxel_shape_spec = (
+            torch.tensor(self._voxel_shape, device=device).unsqueeze(0) + 2
+        )  # +2 because we crop the edges.
+        self._coord_bounds = torch.tensor(
+            coord_bounds, dtype=torch.float, device=device
+        ).unsqueeze(0)
+        max_dims = self._voxel_shape_spec[0]
+        self._total_dims_list = torch.cat(
+            [
+                torch.tensor([batch_size], device=device),
+                max_dims,
+                torch.tensor([4 + feature_size], device=device),
+            ],
+            -1,
+        ).tolist()
+        self._ones_max_coords = torch.ones(
+            (batch_size, max_num_coords, 1), device=device
+        )
+        self._num_coords = max_num_coords
+
+        shape = self._total_dims_list
+        self._result_dim_sizes = torch.tensor(
+            [reduce(mul, shape[i + 1 :], 1) for i in range(len(shape) - 1)] + [1],
+            device=device,
+        )
+        flat_result_size = reduce(mul, shape, 1)
+
+        self._initial_val = torch.tensor(0, dtype=torch.float, device=device)
```

Comment:
very long function - probably best to break it up into multiple sub-functions for clarity